File: SlackOff/Danmaku/DanmakuView.swift
import SwiftUI

struct DanmakuView: View {
    let comments: [DanmakuComment]
    let currentTime: Double
    let settings: DanmakuSettings

    /// 基础滚动速度（字号/速度倍率影响后的 px/s）。
    private var scrollSpeed: CGFloat {
        CGFloat(150 * max(settings.speed, 0.25))
    }

    /// 单条滚动弹幕的轨道行高。
    private var lineHeight: CGFloat {
        max(settings.fontSize + 8, 22)
    }

    var body: some View {
        GeometryReader { geo in
            let laneCount = self.laneCount(size: geo.size)
            let placed = placedComments(size: geo.size, laneCount: laneCount)

            ZStack(alignment: .topLeading) {
                ForEach(placed) { item in
                    DanmakuLabel(
                        comment: item.comment,
                        drift: currentTime - item.comment.time,
                        laneIndex: item.laneIndex,
                        laneCount: laneCount,
                        lineHeight: lineHeight,
                        size: geo.size,
                        settings: settings
                    )
                }
            }
            .clipShape(Rectangle())
        }
        .opacity(settings.opacity)
        .allowsHitTesting(false)
    }

    private struct PlacedItem: Identifiable {
        let id: UUID
        let comment: DanmakuComment
        let laneIndex: Int
    }

    // MARK: - 轨道分配（纯确定性 round-robin，零状态、无跳动、无额外渲染）

    private func placedComments(size: CGSize, laneCount: Int) -> [PlacedItem] {
        let speed = Double(scrollSpeed)
        let maxDrift = Double(size.width) / speed + 8
        let visible = comments
            .filter { $0.time <= currentTime && currentTime < $0.time + maxDrift }
            .sorted { $0.time < $1.time }

        return visible.enumerated().map { index, comment in
            let lane: Int
            if settings.antiOverlap {
                // 防重叠：按时间戳排序后依次分到 0,1,2… 轨道，连续弹幕不撞轨
                lane = index % laneCount
            } else {
                lane = Int(comment.lane) % laneCount
            }
            return PlacedItem(id: comment.id, comment: comment, laneIndex: lane)
        }
    }

    /// 按「显示区域」高度动态决定能塞几行轨道。
    private func laneCount(size: CGSize) -> Int {
        let topInset: CGFloat = 24
        let usable = size.height * settings.region
        return max(1, Int((usable - topInset) / lineHeight))
    }
}

private struct DanmakuLabel: View {
    let comment: DanmakuComment
    let drift: Double
    let laneIndex: Int
    let laneCount: Int
    let lineHeight: CGFloat
    let size: CGSize
    let settings: DanmakuSettings

    var body: some View {
        let speed = Double(150 * max(settings.speed, 0.25))
        let x = size.width + 60 - drift * speed
        Text(comment.text)
            .font(.system(size: settings.fontSize, weight: .semibold))
            .foregroundStyle(textColor)
            .lineLimit(1)
            .fixedSize()
            .position(x: x, y: laneY)
            .opacity(x < -120 ? 0 : 1)
    }

    private var textColor: Color {
        settings.colorPreset.override ?? comment.colorValue
    }

    /// 滚动弹幕：按轨道序号落到固定行高；顶部/底部固定轨道独立。
    private var laneY: CGFloat {
        switch comment.type {
        case 5:
            return 24
        case 4:
            return size.height - 24
        default:
            let band = min(max(laneIndex, 0), max(laneCount - 1, 0))
            let topInset: CGFloat = 24
            return topInset + lineHeight * CGFloat(band) + lineHeight / 2
        }
    }
}

extension DanmakuComment {
    var colorValue: Color {
        let r = Double((color >> 16) & 0xFF) / 255
        let g = Double((color >> 8) & 0xFF) / 255
        let b = Double(color & 0xFF) / 255
        return Color(red: r, green: g, blue: b)
    }
}

File: android/app/src/main/java/com/slackoff/app/danmaku/DanmakuView.kt
package com.slackoff.app.danmaku

import androidx.compose.foundation.layout.*
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shadow
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.slackoff.app.model.DanmakuComment
import kotlin.math.roundToInt

@Composable
fun DanmakuView(
    comments: List<DanmakuComment>,
    currentTime: Double,
    settings: DanmakuSettings,
    topInsetDp: Float = 24f,
    modifier: Modifier = Modifier
) {
    if (!settings.enabled) return
    val density = LocalDensity.current
    // 顶部安全区域偏移（全屏时由调用方传入真实状态栏高度，避免弹幕被刘海/状态栏遮挡）
    val topInsetPx = with(density) { topInsetDp.dp.toPx() }

    val scrollSpeed = (150 * maxOf(settings.speed, 0.25f)).toDouble()
    val lineHeight = with(density) { (maxOf(settings.fontSize, 14f) + 8).sp.toPx() }

    var containerSize by remember { mutableStateOf(androidx.compose.ui.unit.IntSize.Zero) }

    Box(
        modifier = modifier
            .fillMaxSize()
            .alpha(settings.opacity)
            .onSizeChanged { containerSize = it }
    ) {
        if (containerSize.width == 0) return@Box

        val laneCount = maxOf(1, ((containerSize.height * settings.region - topInsetPx) / lineHeight).toInt())
        val maxDrift = containerSize.width / scrollSpeed + 8.0

        val visible = comments
            .filter { it.time <= currentTime && currentTime < it.time + maxDrift }
            .sortedBy { it.time }

        val laneMemory = remember { mutableMapOf<String, Int>() }
        val freeAt = DoubleArray(laneCount) { 0.0 }
        val placed = mutableListOf<Pair<DanmakuComment, Int>>()

        for (comment in visible) {
            val lane = if (laneMemory.containsKey(comment.id)) {
                laneMemory[comment.id]!!
            } else if (settings.antiOverlap) {
                val busy = (60.0 + textWidth(comment.text, settings.fontSize)) / scrollSpeed
                var chosen = -1
                var bestLane = 0
                var bestFree = Double.MAX_VALUE
                for (i in 0 until laneCount) {
                    if (freeAt[i] <= comment.time) { chosen = i; break }
                    if (freeAt[i] < bestFree) { bestFree = freeAt[i]; bestLane = i }
                }
                if (chosen < 0) chosen = bestLane
                laneMemory[comment.id] = chosen
                chosen
            } else {
                (comment.lane.toInt() % laneCount).also { laneMemory[comment.id] = it }
            }
            freeAt[lane] = maxOf(freeAt[lane], comment.time) + (60.0 + textWidth(comment.text, settings.fontSize)) / scrollSpeed
            placed.add(comment to lane)
        }

        for ((comment, laneIndex) in placed) {
            val drift = currentTime - comment.time
            val x = (containerSize.width + 60 - drift * scrollSpeed).toFloat()
            val y = topInsetPx + lineHeight * laneIndex + lineHeight / 2

            if (x < -120) continue

            val textColor = settings.colorPreset.override ?: comment.colorValue

            Box(
                modifier = Modifier
                    .offset { IntOffset(x.roundToInt(), y.roundToInt() - (lineHeight / 2).roundToInt()) }
            ) {
                androidx.compose.material3.Text(
                    text = comment.text,
                    style = TextStyle(
                        fontSize = settings.fontSize.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = textColor,
                        shadow = Shadow(
                            color = Color.Black.copy(alpha = 0.55f),
                            blurRadius = 1.2f,
                            offset = Offset(0f, 1f)
                        )
                    ),
                    maxLines = 1,
                    softWrap = false
                )
            }
        }
    }
}

private val textWidthCache = mutableMapOf<String, Float>()

private fun textWidth(text: String, fontSize: Float): Float {
    val key = "${fontSize.toInt()}|$text"
    return textWidthCache.getOrPut(key) {
        var width = 0f
        for (ch in text) {
            width += if (ch.code > 127) fontSize * 1.0f else fontSize * 0.6f
        }
        width + 8f
    }
}

val DanmakuComment.colorValue: Color
    get() {
        val r = (color shr 16) and 0xFF
        val g = (color shr 8) and 0xFF
        val b = color and 0xFF
        return Color(r, g, b)
    }